server/session.go

```diff
@@ -0,0 +1,194 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	pb "github.com/googleapis/gapic-showcase/server/genproto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Session represents a suite of tests, generally being made in the context
+// of testing code generation.
+//
+// A session defines tests it may expect, based on which version of the
+// code generation spec is in use.
+type Session interface {
+	GetName() string
+	GetVersion() pb.Session_Version
+	GetReport() *pb.ReportSessionResponse
+	RegisterTests(tests []Test)
+	ListTests(in *pb.ListTestsRequest) (*pb.ListTestsResponse, error)
+	DeleteTest(name string) (*empty.Empty, error)
+}
+
+// SessionProto returns a proto representation of the Session.
+func SessionProto(s Session) *pb.Session {
+	return &pb.Session{
+		Name:    s.GetName(),
+		Version: s.GetVersion(),
+	}
+}
+
+// NewSession returns a session for a given name, version and observerRegistry.
+func NewSession(name string, version pb.Session_Version, observerRegistry GrpcObserverRegistry) Session {
+	session := &sessionImpl{
+		name:             name,
+		version:          version,
+		observerRegistry: observerRegistry,
+		token:            NewTokenGenerator(),
+		keys:             map[string]int{},
+		tests:            []testEntry{},
+	}
+	return session
+}
+
+type testEntry struct {
+	test    Test
+	deleted bool
+}
+
+type sessionImpl struct {
+	name             string
+	version          pb.Session_Version
+	observerRegistry GrpcObserverRegistry
+	token            TokenGenerator
+
+	mu    sync.Mutex
+	keys  map[string]int
+	tests []testEntry
+}
+
+func (s *sessionImpl) GetName() string {
+	return s.name
+}
+
+func (s *sessionImpl) GetVersion() pb.Session_Version {
+	return s.version
+}
+
+func (s *sessionImpl) GetReport() *pb.ReportSessionResponse {
+	result := pb.ReportSessionResponse_PASSED
+	for _, entry := range s.tests {
+		if entry.deleted {
+			continue
+		}
+		test := entry.test
+		issue := test.GetIssue()
+		if issue == nil {
+			continue
+		}
+		// TODO: Add severity handling
+		if issue.GetType() == pb.Issue_INCORRECT_CONFIRMATION {
+			result = pb.ReportSessionResponse_FAILED
+			continue
+		}
+		if result != pb.ReportSessionResponse_FAILED {
+			result = pb.ReportSessionResponse_INCOMPLETE
+		}
+	}
+
+	testRuns := []*pb.TestRun{}
+	for _, entry := range s.tests {
+		if entry.deleted {
+			continue
+		}
+		testRuns = append(testRuns, TestRunProto(entry.test))
+	}
+	return &pb.ReportSessionResponse{
+		Result:   result,
+		TestRuns: testRuns,
+	}
+}
+
+func (s *sessionImpl) ListTests(in *pb.ListTestsRequest) (*pb.ListTestsResponse, error) {
+	start, err := s.token.GetIndex(in.GetPageToken())
+	pageSize := in.GetPageSize()
+	if pageSize == 0 {
+		pageSize = 10
+	}
+	if err != nil {
```

Comment:
Nit: the error check should follow the function that emits it

Reply:
Done.